---
title: "01A - Prepare human fetal thalamus data"
author: "Selin Jessa [[selin.jessa@mail.mcgill.ca](mailto:selin.jessa@mail.mcgill.ca)]"
date: "`r format(Sys.time(), '%d %B, %Y')`"
params:
  resources: "NOT SPECIFIED"
output:
  html_document:
    keep_md: yes
    code_folding: show
    theme: flatly
    css: ../include/style.css
    toc: yes
    toc_depth: 3
    number_sections: true
    df_print: paged
    includes: 
      before_body: ../include/header.html
      after_body:  ../include/footer.html
---

<!-- FRONT MATTER, insert configuration info -->

```{r header, echo = FALSE, results = 'asis', warning = FALSE}

# Index of the document
# ...determines name of the subfolder of `outputs` and `figures`
doc_id <- "01A"

suppressMessages(library(here))

# Knit child document with header
res <- knitr::knit_child(here("include", "header.Rmd"), envir = environment(), quiet = TRUE)
cat(res, sep = '\n')

```

<!-- END OF FRONT MATTER -->


# Overview

In this analysis, we'll prepare the reference from the human fetal thalamus data
which has been processed with our single-cell workflow, and gather the
required formats and outputs needed for comparisons to tumors:

* labels from the original publications, as well as manual additions/corrections based on known markers
* Seurat objects, per-sample and integrated across samples
* gene signatures, per-cluster
* mean expression profiles, per-cluster

This data comes from [Bhaduri et al](http://dx.doi.org/10.1038/s41586-021-03910-8), and [Eze et al](http://dx.doi.org/10.1038/s41593-020-00794-1).


# Libraries

```{r libs, warning = FALSE, message = FALSE, echo = TRUE, cache = FALSE}

# Load libraries here
library(here)
library(tidyr)
library(dplyr)
library(readxl)
library(readr)
library(glue)
library(ggplot2)
library(magrittr)
library(purrr)
library(cowplot)
library(pvclust)
library(dendextend)
library(feather)
library(Seurat)
library(icytobox)
library(ggrepel)

source(here("include/style.R"))
source(here("code/functions/scRNAseq.R"))
ggplot2::theme_set(theme_min())

```


# Overview of individual samples

Each sample has a folder in the pipeline at:

```
data/scRNAseq/pipeline_human_fetal/
|-- sample1/
--- |-- preprocessing # output of the SC pipeline
----|-- correlations  # output of projections

```

## Metadata

Load per sample metadata:

```{r info_samples, message = FALSE}

info_samples <- read_tsv(here("data/metadata/metadata_human_fetal_10X.tsv")) %>% 
    filter(Region == "Thalamus") %>% 
    filter(grepl("GW", Age))

palette_sample <- info_samples %>% select(Sample, Sample_color) %>% tibble::deframe()
palette_alias  <- info_samples %>% select(Alias, Sample_color) %>% tibble::deframe()

ages <- info_samples$Age %>% unique %>% sort()
palette_age <- colorRampPalette(brewer.pal(8, "YlGnBu"))(n = length(ages))
names(palette_age) <- ages

```

Load the labels from the authors of the original studies:

```{r published_labels}

metadata_pub_gw <- read_excel(here("data/scRNAseq/references/Bhaduri2021_fetal_brain_GW/GW_sample_labels_Nature_2021_Supp_Table_1.xlsx")) %>% 
    filter(structure == "thalamus")

# get sample names in the same format
samples_pub <- metadata_pub_gw %>% filter(structure == "thalamus") %>%
    pull(cell.name) %>%
    {gsub('.{17}$', '', .)} %>%
    unique()

all(info_samples$Sample_public %in% samples_pub)

```

## QC

Verify QC of each sample, to determine whether to include all:

```{r qc, message = FALSE, fig.width = 5, fig.height = 5, message = FALSE}

qc_stats <- map_dfr(1:nrow(info_samples), ~ read_tsv(here(file.path(info_samples[.x, ]$Path, "preprocessing/seurat_metrics.tsv"))) %>% 
                        tibble::add_column(.before = 1, Sample = info_samples[.x, ]$Sample))

#' a function to plot a scatterplot of 2 column in the aggregated qc dataframe
#' 
#' @example qc_scatter(nCount_RNA_min.postQC, nCount_RNA_mean.postQC)
qc_scatter <- function(col1, col2) {
    
    col1_quo <- enquo(col1)
    col2_quo <- enquo(col2)
    
    qc_stats %>% 
        ggplot(aes(x = !! col1_quo, y = !! col2_quo)) +
        geom_point(aes(color = Sample), size = 4, alpha = 0.5) +
        geom_text_repel(aes(label = Sample, color = Sample), size = 4) +
        scale_color_manual(values = palette_sample) +
        theme_min() +
        no_legend()
    
}

qc_scatter(N_cells_before, N_cells_after)

```

## Prep objects

For each sample:

* load the Seurat object
* count the number of cells in each cluster, and give each cluster a label prefixed w/ sample name
* add the labels from the published metadata

```{r prep_samples, cache = FALSE, eval = TRUE, fig.width = 4.5, height = 4.5, message = FALSE}

prep_seurat <- function(path_to_sample, sample_public_id) {
    
    message("@ ", basename(path_to_sample))
    
    # load the object as output by the SC workflow
    load(file.path(path_to_sample, "preprocessing/seurat.Rda"))
    
    # get the labels from the published metadata and add them to the Seurat object
    public_labels <- seurat@meta.data %>%
        tibble::rownames_to_column(var = "cell.name") %>%
        rowwise() %>%
        mutate(barcode_public = paste0(sample_public_id, "_", gsub("-1", "", cell.name, fixed = TRUE))) %>%
        left_join(metadata_pub_gw %>%
                      select(barcode_public = cell.name, Label_public = `cell type`), by = c("barcode_public")) %>% 
        select(cell.name, Label_public) %>%
        tibble::column_to_rownames(var = "cell.name")
    
    seurat <- AddMetaData(seurat,
                          metadata = public_labels,
                          col.name = c("Label_published"))
    
    # get some summary stats per cluster
    seurat[["Cluster_number"]] <- Idents(object = seurat)
    
    info_samples_cluster <- seurat@meta.data %>%
        group_by(Cluster_number) %>%
        mutate(N_cells = n()) %>%
        group_by(Cluster_number, N_cells) %>%
        summarize_at(vars(nCount_RNA, nFeature_RNA, percent.mito, percent.ribo, S.Score, G2M.Score), mean) %>%
        tibble::add_column(.before = 1, Sample = basename(path_to_sample)) %>%
        # add sample info table, and prefix the cluster number by the alias, so that it's unique
        # in the form, [Alias]_[Cluster]
        left_join(select(info_samples, Sample, Age, Alias), by = "Sample") %>%
        mutate(Cluster = paste0(Alias, "_", Cluster_number)) %>%
        select(Sample, Age, Alias, Cluster_number, Cluster, everything())
    
    # get the most common projection from the public data
    cluster_labels_published <- seurat@meta.data %>% rowwise() %>% 
        group_by(Cluster_number) %>% 
        mutate(N_cells_in_cluster = n()) %>% 
        filter(!is.na(Label_published)) %>% 
        group_by(Cluster_number, N_cells_in_cluster, Label_published) %>%
        summarize(N_celltype_in_cluster = n()) %>% 
        mutate(Prop_celltype_in_cluster = round(N_celltype_in_cluster/N_cells_in_cluster, 2)) %>%
        group_by(Cluster_number) %>% top_n(1, Prop_celltype_in_cluster) %>%
        slice(1) %>%
        filter(Prop_celltype_in_cluster >= 0.25) %>% 
        select(Cluster_number, N_celltype_in_cluster, Prop_celltype_in_cluster, Label_published)
    
    info_samples_cluster <- info_samples_cluster %>%
        left_join(cluster_labels_published, by = "Cluster_number") %>%
        mutate(ID_20220321 = paste0(Cluster, "_", Label_published)) %>% 
        dplyr::rename(Prop_cluster_with_published_label = Prop_celltype_in_cluster) %>% 
        select(-N_celltype_in_cluster)
    
    return(info_samples_cluster)
    
}

# gather all cluster summary info
info_clusters <- map_dfr(1:nrow(info_samples), ~ prep_seurat(info_samples[.x, ]$Path, info_samples[.x, ]$Sample_public))
write_tsv(info_clusters, glue("{out}/info_clusters.tsv"))

```

```{r read_ic, message = FALSE}

info_clusters <- read_tsv(glue("{out}/info_clusters.tsv"))

```

## Sample exclusions

We'll exclude the following samples:

* GW19_thalamus, due to low # of cells
* GW14_thalamus, due to low structure
* GW18_2_ventral_thalamus, due to high mito content
* GW20_34_dorsalthalamus, due to low coverage

## Per-cluster QC

```{r clusters_qc}

samples_drop <- c("GW19_thalamus", "GW14_thalamus", "GW18_2_ventral_thalamus", "GW20_34_dorsalthalamus")

info_clusters %>% nrow()

info_clusters2 <- info_clusters %>% 
    mutate(Note_sample = case_when(
        Sample == "GW19_thalamus" ~ "Sample excluded, low number of cells",
        Sample == "GW14_thalamus" ~ "Sample excluded, low structure",
        Sample == "GW18_2_ventral_thalamus" ~ "Sample excluded, high mitochondrial content",
        Sample == "GW20_34_dorsalthalamus" ~ "Sample excluded, low coverage",
        TRUE ~ "OK"),
        Note_cluster = case_when(
            Note_sample != "OK" ~ "Sample excluded",
            N_cells < 30 ~ "Less than 30 cells",
            nCount_RNA < 1000 ~ "Low number of UMIs",
            percent.ribo > 50 ~ "High ribosomal content",
            TRUE ~ "OK"
        ))

info_clusters2 %>% filter(Note_sample == "OK" & Note_cluster == "OK") %>% nrow()
write_tsv(info_clusters2, glue("{out}/info_clusters2.tsv"))

# total # of cells
sum(info_clusters2$N_cells)

```

## QC tables

Produce a QC table for the supplementary materials:

```{r qc_table}

TABLE_thalamus_qc <- info_samples %>% 
    left_join(qc_stats, by = "Sample") %>% 
    mutate(Note_sample = case_when(
        Sample == "GW19_thalamus" ~ "Sample excluded, low number of cells",
        Sample == "GW14_thalamus" ~ "Sample excluded, low structure",
        Sample == "GW18_2_ventral_thalamus" ~ "Sample excluded, high mitochondrial content",
        Sample == "GW20_34_dorsalthalamus" ~ "Sample excluded, low coverage",
        TRUE ~ "OK")) %>% 
    mutate(Publication = case_when(
        grepl("CS", Sample) ~ "Eze et al, Nature Neuroscience, 2021",
        grepl("GW", Sample) ~ "Bhaduri et al, Nature, 2021"
    )) %>% 
    select(
        # Sample info
        Sample, Age, Alias, Region, Publication,
        # QC
        # N_reads = `Number of Reads`,
        # N_cells_estimated = `Estimated Number of Cells`,
        N_cells_after_filtering = N_cells_after,
        # Reads_mapped_to_genome = `Reads Mapped to Genome`,
        # Reads_mapped_to_transcriptome = `Reads Mapped Confidently to Transcriptome`,
        Mean_mitochondrial_content_after_filtering = percent.mito_mean.postQC,
        Mean_UMIs_after_filtering = nCount_RNA_mean.postQC,
        Mean_N_genes_after_filtering = nFeature_RNA_mean.postQC,
        Max_mito_threshold = percent.mito_max.threshold,
        Min_N_genes_threshold = nFeature_RNA_min.threshold,
        Max_N_genes_threshold = nFeature_RNA_max.threshold,
        Max_UMIs_threshold = nCount_RNA_max.threshold,
        Note_sample)

rr_write_tsv(TABLE_thalamus_qc,
             glue("{out}/TABLE_thalamus_QC.tsv"),
             "Summary of sample info and QC for thalamus samples")

```

# Gather gene signatures

First, load all cluster markers and filter out markers for samples/clusters that we'll exclude:

```{r generate_signatures, message = FALSE}

clusters_keep <- info_clusters2 %>% filter(Note_cluster == "OK") %>% pull(Cluster)

cluster_markers <- map_dfr(1:nrow(info_samples), ~ read_tsv(here(file.path(info_samples[.x, ]$Path, "preprocessing/cluster_markers.tsv"))) %>% 
                               tibble::add_column(Sample = info_samples[.x, ]$Sample)) %>% 
    # get positive markers
    filter(avg_logFC > 0) %>%
    mutate(cluster = as.numeric(cluster)) %>% 
    # add the cluster/sample info
    left_join(info_clusters2 %>% select(Sample, Cluster_number, Cluster) %>% mutate(Cluster_number = as.numeric(as.character(Cluster_number))),
              by = c("Sample" = "Sample", "cluster" = "Cluster_number")) %>% 
    rename(Cluster_number = cluster) %>% 
    # filter out excluded samples and clusters
    filter(Cluster %in% clusters_keep) %>% 
    # get the diff between pct.1 and pct.2
    mutate(pct_diff = pct.1 - pct.2) %>% 
    # reorder
    select(gene, pct_diff, everything())

# sanity checks
unique(cluster_markers$Cluster) %>% length()
any(cluster_markers$Sample %in% samples_drop)
all(cluster_markers$Cluster %in% clusters_keep)

head(cluster_markers)

# save
write_tsv(cluster_markers, glue("{out}/cluster_markers.tsv"))

```

We'll also want to remove erythrocyte & melanocyte clusters, which are quite biologically distinct:

```{r filter_erythro_melano}

erythro_markers <- c("HBB", "HBA1", "HBA2", "HBZ", "HBG1")
(erythro_count <- cluster_markers %>% filter(gene %in% erythro_markers) %>% pull(Cluster) %>% table())

# remove the clusters with 4-5 of these markers:
erythro_clusters <- names(erythro_count)[erythro_count >= 4]

melano_markers <- c("DCT", "MITF", "PMEL")
(melano_count <- cluster_markers %>% filter(gene %in% melano_markers) %>% pull(Cluster) %>% table())

# remove the clusters with all three of these:
melano_clusters <- names(melano_count)[melano_count == 3]

```

# Complete and refine cluster labels

We will manually update the cluster labels (provided by the author) in order to ensure all populations 
of interest are labelled. We will use the following markers:

* Astrocytes: FABP7, S100B, CLU, AQP4
* Neurons: STMN2, TUBB, SYT1
* Microglia: C1QC, LY86
* Ependymal: FOXJ1,  DNAH10, KIF19, DNAH12, TEK1
* OPC: PDGFRA, OLIG1, OLIG2
* Glial progenitors: SOX2, OLIG1/2

After using these markers to update the labels outside of R, we load in the updated labels:

```{r manual_labels, message = FALSE}

labels_manual <- read_tsv(here("data/scRNAseq/references/Bhaduri2021_fetal_brain_GW/2022-03-21-human_fetal_thalamus_labels_manual_SJ.tsv")) %>% 
    select(Cluster, Label_manual)

info_clusters3 <- info_clusters2 %>% 
    left_join(labels_manual) %>% 
    mutate(Note_cluster = case_when(
        grepl("Outlier", Label_published) ~ "Exclude cluster, outlier",
        Label_manual == "Erythrocytes" ~ "Exclude cluster, erythrocytes",
        Label_manual == "Melanocytes" ~ "Exclude cluster, melanocytes",
        TRUE ~ Note_cluster
    )) %>% 
    mutate(ID_20220321 = case_when(
        Note_sample != "OK" | Note_cluster != "OK" ~ paste0(Cluster, "_EXCLUDE"),
        is.na(Label_manual) ~ ID_20220321,
        !is.na(Label_manual) ~ paste0(Cluster, "_", Label_manual)
    )) %>% 
    select(-Label_manual)

write_tsv(info_clusters3, glue("{out}/info_clusters3.tsv"))

clusters_keep <- info_clusters3 %>% filter(!grepl("EXCLUDE", ID_20220321)) %>% pull(Cluster)
length(clusters_keep)    

```


# Make reference for cell type projection

In this section, we prepare the reference matrix for cell type projection,
containing the mean expression of each gene in each cluster.

## Load data

```{r load_int, cache = FALSE}

load(here("data/scRNAseq/integrations/human_fetal_thalamus/output/seurat_joint.Rda"))

```


## Mean expression

Calculate mean expression per cluster:

```{r calculate_meanexp}

Idents(seurat_joint) <- "Cluster"
hf_thalamus_meanexp <- AverageExpression(seurat_joint, return.seurat = TRUE, assays = "RNA", verbose = FALSE)
hf_thalamus_meanexp <- GetAssayData(hf_thalamus_meanexp)

# peek at the corner
dim(hf_thalamus_meanexp)
hf_thalamus_meanexp[1:5, 1:5]

hf_thalamus_meanexp <- hf_thalamus_meanexp[, clusters_keep]

# sanity checks
dim(hf_thalamus_meanexp)
length(clusters_keep)

save(hf_thalamus_meanexp, file = glue("{out}/hf_thalamus_mean_expression.Rda"))
saveRDS(hf_thalamus_meanexp, glue("{out}/hf_thalamus_mean_expression.Rds"))

```

## Annotation

Make the annotation for the cell type projection scripts:

```{r write_anno}

info_clusters3 %>%
    filter(Cluster %in% clusters_keep) %>% 
    select(Cell_type = Cluster, ID_20220321) %>% 
    mutate(Color = case_when(
        grepl("Proliferating OPC", ID_20220321) ~ palette_type[["Proliferating OPC"]],
        grepl("OPC", ID_20220321) ~ palette_type[["OPC"]],
        grepl("Oligo", ID_20220321) ~ palette_type[["Oligodendrocytes"]],
        grepl("[Nn]euron", ID_20220321) ~ palette_type[["Neurons"]],
        grepl("Astro", ID_20220321) ~ palette_type[["Astrocytes"]],
        grepl("Ependymal", ID_20220321) ~ palette_type[["Ependymal"]],
        grepl("Micro", ID_20220321) ~ palette_type[["Immune"]],
        grepl("Endo", ID_20220321) ~ palette_type[["Vascular & other"]],
        grepl("Dividing|RG", ID_20220321) ~ palette_type[["RGC"]],
    )) %>% 
    select(-ID_20220321) %>%
    write_tsv(glue("{out}/hf_thalamus_annotation.tsv"))

```

<!-- END MATTER, insert reproducibility info -->

```{r footer, echo = FALSE, results = 'asis', warning = FALSE, cache = FALSE}

# Knit child document with header
res <- knitr::knit_child(here("include", "footer.Rmd"), envir = environment(), quiet = TRUE)
cat(res, sep = '\n')

```

<!-- END OF END MATTER -->
